stdlib/public/core/KeyValuePairs.swift

```diff
@@ -0,0 +1,141 @@
+//===--- KeyValuePairs.swift ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A lightweight collection of key-value pairs.
+///
+/// Use a `KeyValuePairs` instance when you need an ordered collection of
+/// key-value pairs and don't require the fast key lookup that the
+/// `Dictionary` type provides. Unlike key-value pairs in a true dictionary,
+/// neither the key nor the value of a `KeyValuePairs` instance must
+/// conform to the `Hashable` protocol.
+///
+/// You initialize a `KeyValuePairs` instance using a Swift dictionary
+/// literal. Besides maintaining the order of the original dictionary literal,
+/// `KeyValuePairs` also allows duplicates keys. For example:
+///
+///     let recordTimes: KeyValuePairs = ["Florence Griffith-Joyner": 10.49,
+///                                           "Evelyn Ashford": 10.76,
+///                                           "Evelyn Ashford": 10.79,
+///                                           "Marlies Gohr": 10.81]
+///     print(recordTimes.first!)
+///     // Prints "("Florence Griffith-Joyner", 10.49)"
+///
+/// Some operations that are efficient on a dictionary are slower when using
+/// `KeyValuePairs`. In particular, to find the value matching a key, you
+/// must search through every element of the collection. The call to
+/// `firstIndex(where:)` in the following example must traverse the whole
+/// collection to find the element that matches the predicate:
+///
+///     let runner = "Marlies Gohr"
+///     if let index = recordTimes.firstIndex(where: { $0.0 == runner }) {
+///         let time = recordTimes[index].1
+///         print("\(runner) set a 100m record of \(time) seconds.")
+///     } else {
+///         print("\(runner) couldn't be found in the records.")
+///     }
+///     // Prints "Marlies Gohr set a 100m record of 10.81 seconds."
+///
+/// Dictionary Literals as Function Parameters
```

Comment:
I *think* this should read "Key-Value Pairs as Function Parameters", as the text talks about using `KeyValuePair` parameters with dictionary literals as arguments.

Reply:
Good spot. I'll do that in another PR and land this tho.